include/openthread/commissioner.h

```diff
@@ -112,19 +113,42 @@ typedef struct otCommissioningDataset
     bool mIsJoinerUdpPortSet : 1; ///< TRUE if Joiner UDP Port is set, FALSE otherwise.
 } otCommissioningDataset;
 
-#define OT_PSKD_MAX_SIZE 32 ///< Size of a Joiner PSKd (bytes)
+#define OT_JOINER_MAX_PSKD_LENGTH 32 ///< Maximum string length of a Joiner PSKd (does not include null char).
+
+/**
+ * This structure represent a Joiner PSKd.
+ *
+ */
+typedef struct otJoinerPskd
+{
+    char m8[OT_JOINER_MAX_PSKD_LENGTH + 1]; ///< Char string array (must be null terminated - +1 is for null char).
+} otJoinerPskd;
+
+/**
+ * This enumeration defines a Joiner Info Typer.
+ *
+ */
+typedef enum otJoinerInfoType
+{
+    OT_JOINER_INFO_TYPE_ANY       = 0, ///< Accept any Joiner (no EUI64 or Discerner is specified).
+    OT_JOINER_INFO_TYPE_EUI64     = 1, ///< Joiner EUI-64 is specified (`mShared.mEui64` in `otJoinerInfo`).
+    OT_JOINER_INFO_TYPE_DISCERNER = 2, ///< Joiner Discerner is specified (`mShared.mDiscerner` in `otJoinerInfo`).
+} otJoinerInfoType;
 
 /**
  * This structure represents a Joiner Info.
  *
  */
 typedef struct otJoinerInfo
 {
-    otExtAddress mEui64;                     ///< Joiner eui64
-    char         mPsk[OT_PSKD_MAX_SIZE + 1]; ///< Joiner pskd
-    uint32_t     mExpirationTime;            ///< Joiner expiration time in msec
-
-    bool mAny : 1; /// TRUE if eui64 isn't set, FALSE otherwise.
+    otJoinerInfoType mType; ///< Joiner type.
+    union
+    {
+        otExtAddress      mEui64;     ///< Joiner EUI64 (when `mType` is `OT_JOINER_INFO_TYPE_EUI64`)
+        otJoinerDiscerner mDiscerner; ///< Joiner Discerner (when `mType` is `OT_JOINER_INFO_TYPE_DISCERNER`)
+    } mShared;                        ///< Shared fields
```

Comment:
We use `mShared` to indicate `union` fields.  I used "Joiner Discerner" and avoided the  "discriminator" term to ensure no confusion with CHIP project use of this term (also it has the side benefit of not using ["discriminate"](https://en.wikipedia.org/wiki/Discrimination) term  which is generally associated with not-positive stuff).. :)

Reply:
`mEui64` and `mDiscerner` are both objects represent an identity, so I was thinking that the union should be a common / abstract name describes the meaning of the two fields. While discriminator may not be a good name,  what about `mSharedId` we used in `commissioner.h`?

Reply:
Another question is, do we need to distinguish `EUI64` and general `discerner`? I don't see how we will process differently for EUI64 and discerner, since they are just sources of joiner ID, which is used for steering data. It seems we can remove `mType` and `mEui64` and keep only `mDiscerner`. Thoughts?

Reply:
I am ok renaming the `union` field to `mSharedId`. Will do in the next push.

I think we do need to distinguish between EUI64 and discerner (I think you may be thinking that we can treat EUI64 as a discerner with 64 bit length which would not be the case). 
- In terms of how they are used there are differences: 
   - For EUI64 we use a hash of it to add to bloom filter and also use same as Joiner Id.
   - For Joiner Discerner, there is no hashing for bloom filter (the actual discerner value is directly used for updating bloom filter). For link-local IID (Joiner Id) we currently do add a random prefix to extend it to 64 bit (if not already at 64) and no hashing. We may consider later to update this, e.g., use hash of a random prefix + discerner value to hide the actual discerner value when deriving the link-local IID (note that this would be only for IID derivation and cannnot be used in bloom filter).
- Also in terms of representation (in `JoinerInfo`),  the discerner is an unsigned int `uint64_t` value so its byte order is machine/platform dependent (little/big endian), compared to `EUI64` which is a MAC Ext Address (byte order is fixed).

Reply:
> For Joiner Discerner, there is no hashing for bloom filter

I missed this. Agree we need to keep both `EUI64` and `Discerner`.